lib/CL/pocl_llvm_wg.cc

```diff
@@ -160,39 +155,227 @@ static TargetMachine *GetTargetMachine(cl_device_id device, Triple &triple) {
+  OptimizeLevel = OLevel;
+  SizeLevel = SLevel;
 
-  // Need to setup the target info for target specific passes. */
-  Triple triple(device->llvm_target_triplet);
-  TargetMachine *Machine = GetTargetMachine(device, triple);
+#if 0
+  PrintPassOpts.Verbose = true;
+  PrintPassOpts.SkipAnalyses = false;
+  PrintPassOpts.Indent = true;
+  SI.reset(new StandardInstrumentations(Context,
+                                        true, // debug logging
+                                        false, // verify each
+                                        PrintPassOpts));
+  SI->registerCallbacks(PIC, &MAM);
+#endif
+  // Create the new pass manager builder.
+  // Take a look at the PassBuilder constructor parameters for more
+  // customization, e.g. specifying a TargetMachine or various debugging
+  // options.
+  // PassB.reset(new PassBuilder(Machine, PTO, std::nullopt, &PIC));
+  PassB.reset(new PassBuilder(Machine, PTO));
+  PassBuilder &PB = *PassB.get();
+
+  // Register our TargetLibraryInfoImpl.
+  TargetLibraryInfoImpl TLII(DevTriple);
+  // Disables automated generation of libcalls from code patterns.
+  // TCE doesn't have a runtime linker which could link the libs later on.
+  // Also the libcalls might be harmful for WG autovectorization where we
+  // want to try to vectorize the code it converts to e.g. a memset or
+  // a memcpy
+  TLII.disableAllFunctions();
+  // Analysis pass providing the \c TargetLibraryInfo:
+  // TargetLibraryAnalysis
 
-  if (Machine)
-    Passes->add(
-        createTargetTransformInfoWrapperPass(Machine->getTargetIRAnalysis()));
+  pocl::registerFunctionAnalyses(PB);
+  // early register here, this will automatically skip later registrations
+  FAM.registerPass([&] { return TargetLibraryAnalysis(TLII); });
 
+  // Register all the basic analyses with the managers.
+  PB.registerModuleAnalyses(MAM);
+  PB.registerCGSCCAnalyses(CGAM);
+  PB.registerFunctionAnalyses(FAM);
+  PB.registerLoopAnalyses(LAM);
+  PB.crossRegisterProxies(LAM, FAM, CGAM, MAM);
 
-  /* Disables automated generation of libcalls from code patterns.
-     TCE doesn't have a runtime linker which could link the libs later on.
-     Also the libcalls might be harmful for WG autovectorization where we
-     want to try to vectorize the code it converts to e.g. a memset or
-     a memcpy */
-  TargetLibraryInfoImpl TLII(triple);
-  TLII.disableAllFunctions();
-  Passes->add(new TargetLibraryInfoWrapperPass(TLII));
+  pocl::registerPassBuilderPasses(PB);
 
+  return PB.parsePassPipeline(PM, StringRef(PoclPipeline));
+}
+
+void PoclModulePassManager::run(llvm::Module &Bitcode) {
+  PM.run(Bitcode, MAM);
+  populateModulePM(nullptr, (void *)&Bitcode, OptimizeLevel, SizeLevel,
+                   Vectorize);
+}
+
+class TwoStagePoclModulePassManager {
+  PoclModulePassManager Stage1;
+  PoclModulePassManager Stage2;
+
+public:
+  TwoStagePoclModulePassManager() = default;
+  llvm::Error build(cl_device_id Dev, const std::string &S1_Pipeline,
```

Comment:
Check the mixed naming style in the variables, let's try to aim for LLVM C++ style in new C++ codes committed in.